exonum/src/blockchain/block.rs

```diff
@@ -12,15 +12,85 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use failure::Error;
+
 use exonum_merkledb::MapProof;
 use exonum_proto::ProtobufConvert;
 
 use crate::{
     crypto::Hash,
     helpers::{Height, ValidatorId},
     messages::{Precommit, Verified},
-    proto,
+    proto::{self, BinaryMap},
 };
+use exonum_merkledb::BinaryValue;
+use std::{borrow::Cow, fmt};
+
+/// Trait that represents key in block header entry map. Provide
+/// mapping between `NAME` of the entry and its value.
+///
+/// # Usage
+///
+/// see [`Block::get_entry()`].
+///
+/// [`Block::get_entry()`]: struct.Block.html#method.get_entry
+pub trait BlockHeaderKey {
+    /// Key name.
+    const NAME: &'static str;
+    /// Type of the value associated with this key.
+    type Value: BinaryValue;
+}
+
+/// Proposer identifier.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct ProposerId(pub ValidatorId);
+
+impl fmt::Display for ProposerId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl BinaryValue for ProposerId {
+    fn to_bytes(&self) -> Vec<u8> {
+        (self.0).0.to_bytes()
+    }
+
+    fn from_bytes(bytes: Cow<'_, [u8]>) -> Result<Self, Error> {
+        Ok(Self(ValidatorId(u16::from_bytes(bytes)?)))
+    }
+}
+
+impl From<ValidatorId> for ProposerId {
+    fn from(validator_id: ValidatorId) -> Self {
+        ProposerId(validator_id)
+    }
+}
+
+impl BlockHeaderKey for ProposerId {
+    const NAME: &'static str = "proposer_id";
+    type Value = Self;
```

Comment:
Why not use `type Value = ValidatorId`? Would allow to skip conversions elsewhere. In this case `ProposerId` could be an empty type.